### PR TITLE
Feat/widget conf persist

### PR DESCRIPTION
**Pull request by amevck**

### CODE REVIEW

File: src/components/WidgetsCustomizer/WidgetsCustomizer.tsx [hunk 1]
```diff
@@ -27,6 +27,9 @@ import './WidgetsCustomizer.css';
 
 import WIDGET_SETTINGS from 'constants/widgetSettings';
 import { WidgetConfigProps } from 'components/widgetConfigs/event/interfaces';
+import store from 'store';
+import { SET_NEW_WIDGET } from 'store/actions/actionTypes';
```
Comment (man-csl) [inline src/components/WidgetsCustomizer/WidgetsCustomizer.tsx:31]: SET_NEW_WIDGET -> does this reflect the purpose of this action?

File: src/components/WidgetsCustomizer/WidgetsCustomizer.tsx [hunk 2]
```diff
@@ -143,8 +158,8 @@ export default function WidgetsCustomizer() {
   useEffect(() => {
     updateSizeMapping();
   }, []); // eslint-disable-line react-hooks/exhaustive-deps
-  const Configurator: FC<WidgetConfigProps> =
-    WIDGET_SETTINGS[selectedWidgetKey].configurator;
+  const Configurator: FC<WidgetConfigProps> = getWidgetSettingKey()
```
Comment (man-csl) [inline src/components/WidgetsCustomizer/WidgetsCustomizer.tsx]: why is this called getWidgetSettingKey? But looks like it returns all settings?

File: src/components/grid/GridContainer.tsx [hunk 3]
```diff
@@ -1,105 +1,167 @@
...
-        return newLayout;
-      });
-      return;
-    }
+  /**
+   * compare 2 layouts position is same or not
+   * @param layOut1
+   * @param layOut2
+   */
+  const onError = (msg: string) => {
```
Comment (man-csl) [inline src/components/grid/GridContainer.tsx]: Is this comment correct?

File: src/components/grid/GridContainer.tsx [hunk 4]
```diff
@@ -1,105 +1,167 @@
 // Copyright 2020 Cognite AS
-import React, { FC, useState } from 'react';
+import React, { FC, useState, useEffect, useRef } from 'react';
```
Comment (man-csl) [inline src/components/grid/GridContainer.tsx:2]: Need to understand why these changes are required for config save thingy..

File: src/components/WidgetsCustomizer/WidgetsCustomizer.tsx [hunk 5]
```diff
@@ -119,8 +123,19 @@ export default function WidgetsCustomizer() {
       }
     }
   };
-  const onCreate = (data: any) => {
-    console.log('on parent', data);
+  const getWidgetSettingKey = (objKey: string | null = null) => {
```
Comment (man-csl) [inline src/components/WidgetsCustomizer/WidgetsCustomizer.tsx]: for me it looks like this returns the widget settings? not the key? Does the name of the function reflect that?

File: src/components/WidgetsCustomizer/WidgetsCustomizer.tsx [hunk 6]
```diff
@@ -119,8 +123,19 @@ export default function WidgetsCustomizer() {
...
+    if (!objKey) {
+      return WIDGET_SETTINGS[selectedWidgetKey];
```
Comment (man-csl) [inline src/components/WidgetsCustomizer/WidgetsCustomizer.tsx]: Why cant the caller make this decision?

File: src/components/grid/GridContainer.tsx [hunk 7]
```diff
@@ -1,105 +1,171 @@
...
-      prevLayout.filter((compDetails: Layout) => compDetails.i !== key)
-    );
-    setWidgetConfigs((prevWidgetConfigs: WidgetConfig[]) =>
-      prevWidgetConfigs.filter(
-        (compDetails: WidgetConfig) => compDetails.i !== key
-      )
+  const onLayoutChange = (newLayouts: Layout[]) => {
+    setLayouts(newLayouts);
+  };
+  const isOnSettingPage = useLocation().pathname === RouterPaths.SETTINGS;
```
Comment (man-csl) [inline src/components/grid/GridContainer.tsx:50]: can this be injected?

File: src/components/grid/GridContainer.tsx [hunk 8]
```diff
@@ -1,105 +1,171 @@
...
+  /**
+   * fire when a widget is moved and place in a different position
```
Comment (man-csl) [inline src/components/grid/GridContainer.tsx]: place -> placed

File: src/components/grid/GridContainer.tsx [hunk 9]
```diff
@@ -1,105 +1,171 @@
...
-          i: key,
-          widgetTypeId: WIDGET_TYPE_IDS.ASSET_INFO,
-          valueMapping,
-        });
-        return prevWidgetConfigs;
-      });
+  const isMounted = useRef(true);
+  const [widgetConfigs, setWidgetConfigs] = useState<WidgetConfig[]>([]);
+  const [layouts, setLayouts] = useState<Layout[]>([]);
+  const [lastSavedLayouts, setLastSavedLayouts] = useState<Layout[]>([]);
```
Comment (man-csl) [inline src/components/grid/GridContainer.tsx:36]: Why do you need to keep an array here? What does a Layout represent?

File: src/components/grid/GridContainer.tsx [hunk 10]
```diff
@@ -1,105 +1,171 @@
...
+    if (isSuccess) {
+      setLayouts((prevLayout: Layout[]) =>
+        prevLayout.filter((compDetails: Layout) => compDetails.i !== widgetId)
+      );
+      setWidgetConfigs((prevWidgetConfigs: WidgetConfig[]) =>
+        prevWidgetConfigs.filter(
+          (compDetails: WidgetConfig) => compDetails.i !== widgetId
+        )
+      );
+    }
```
Comment (man-csl) [inline src/components/grid/GridContainer.tsx:121]: I thought layout stores information about locations of widgets. Widget confic stores configurations of each widget such as name of time series, etc. In that case why do we have to change widget config when one is removed from the layout.

File: src/components/grid/GridContainer.tsx [hunk 11]
```diff
@@ -1,105 +1,171 @@
...
+    console.log(layouts, widgetCordinates);
+    if (!widgetCordinates) {
+      onError('There is no position for adding the component');
+      return;
+    }
+    const newWidgetConf = { ...widgetConfig };
+    newWidgetConf.i = generateRandomKey();
+    newWidgetConf.cordinates = widgetCordinates;
+    const newWidgetConfs = [...widgetConfigs].concat(newWidgetConf);
+    const isSuccess = saveWidget(
```
Comment (man-csl) [inline src/components/grid/GridContainer.tsx]: saveWidget -> saveWidgetConfig?

File: src/components/grid/GridLayout/GridLayout.tsx [hunk 12]
```diff
@@ -1,10 +1,9 @@
 // Copyright 2020 Cognite AS
```
Comment (man-csl) [inline src/components/grid/GridLayout/GridLayout.tsx:1]: What does GridContainer and GridLayout represent ceonceptually? Can you add comments explaining that?

File: src/components/widgetCRUD/repo/widgstConfFSRepo.ts [hunk 13]
```diff
@@ -0,0 +1,70 @@
+// Copyright 2020 Cognite AS
```
Comment (man-csl) [inline src/components/widgetCRUD/repo/widgstConfFSRepo.ts:1]: please add comments to this and previous one explaining the difference.

File: src/components/widgetCRUD/services/widgetConfService.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
+// Copyright 2020 Cognite AS
```
Comment (man-csl) [inline src/components/widgetCRUD/services/widgetConfService.ts:1]: More comments explaining the purpose of this code please.

File: src/components/widgetCRUD/services/widgetConfService.ts [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+import { Layout } from 'react-grid-layout';
+import findIndex from 'lodash/findIndex';
+import { WidgetConfig } from 'components/grid/interfaces';
+import { getByUser, update, deleteOne, save } from '../repo/widgstConfFSRepo';
```
Comment (man-csl) [inline src/components/widgetCRUD/services/widgetConfService.ts:5]: Isn't there a better way to inject the repo as an interface so that the service does not have to hardcode the actual implementation it uses?

File: src/firebase.ts [hunk 16]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2020 Cognite AS
+import firebase from 'firebase';
+import { WIDGET_CONF_DB } from 'constants/widgetSettings';
+
+const firebaseConfig = {
+  apiKey: process.env.REACT_APP_API_KEY,
```
Comment (man-csl) [inline src/firebase.ts:6]: Since this is loaded to client browser, the app key gets exposed. How do we make sure that someone having access to this key will not be able to access the firebase store?

File: src/components/WidgetsCustomizer/WidgetsCustomizer.tsx (hunk 1 above)
Comment (amevck) [reply src/components/WidgetsCustomizer/WidgetsCustomizer.tsx:31]: I think so. because this is to fire when we want to add a new widget.

File: src/components/WidgetsCustomizer/WidgetsCustomizer.tsx (hunk 2 above)
Comment (amevck) [reply src/components/WidgetsCustomizer/WidgetsCustomizer.tsx]: fixed

File: src/components/grid/GridContainer.tsx (hunk 3 above)
Comment (amevck) [reply src/components/grid/GridContainer.tsx]: it should be removed. it's relevant to a method that is already removed.

File: src/components/grid/GridContainer.tsx (hunk 4 above)
Comment (amevck) [reply src/components/grid/GridContainer.tsx:2]: This is not relevant to the saving process. but to dom updating with react. after save/update/delete we have to use these effects to update them correctly.

File: src/components/WidgetsCustomizer/WidgetsCustomizer.tsx (hunk 5 above)
Comment (amevck) [reply src/components/WidgetsCustomizer/WidgetsCustomizer.tsx]: fixed with an above comment

File: src/components/WidgetsCustomizer/WidgetsCustomizer.tsx (hunk 6 above)
Comment (amevck) [reply src/components/WidgetsCustomizer/WidgetsCustomizer.tsx]: method removed

File: src/components/grid/GridContainer.tsx (hunk 7 above)
Comment (amevck) [reply src/components/grid/GridContainer.tsx:50]: No. this is a hook. this acts the same as injected. As discussed in a previous PR this is hard to pass over component hierarchy to the child.

File: src/components/grid/GridContainer.tsx (hunk 8 above)
Comment (amevck) [reply src/components/grid/GridContainer.tsx]: done

File: src/components/grid/GridContainer.tsx (hunk 9 above)
Comment (amevck) [reply src/components/grid/GridContainer.tsx:36]: This is to remove in #86. however this is for smooth the saving process. and when a failure happens this is useful to revert back the changes.

File: src/components/grid/GridContainer.tsx (hunk 10 above)
Comment (amevck) [reply src/components/grid/GridContainer.tsx:121]: We have to remove both widget configs and layout when a widget removed. (these widget configs are not statics these are object-level dynamic configurations) not just the location but also configurations. However, I'll remove these layout and widget config operate separation in the next PRs.

File: src/components/grid/GridContainer.tsx (hunk 11 above)
Comment (amevck) [reply src/components/grid/GridContainer.tsx]: done

File: src/components/grid/GridLayout/GridLayout.tsx (hunk 12 above)
Comment (amevck) [reply src/components/grid/GridLayout/GridLayout.tsx:1]: GridLayout only a wrapper component to the react-grid-layout library (leaf component). it has things like styling, managing resizing. but the grid container has all capabilities such as manage state(data), handle side effects, etc.

File: src/components/widgetCRUD/repo/widgstConfFSRepo.ts (hunk 13 above)
Comment (amevck) [reply src/components/widgetCRUD/repo/widgstConfFSRepo.ts:1]: done

File: src/components/widgetCRUD/services/widgetConfService.ts (hunk 14 above)
Comment (amevck) [reply src/components/widgetCRUD/services/widgetConfService.ts:1]: there will be some changes in the PR #86. I'll add the comments there because some conflicts may occur if I changed this now.

File: src/components/widgetCRUD/services/widgetConfService.ts (hunk 15 above)
Comment (amevck) [reply src/components/widgetCRUD/services/widgetConfService.ts:5]: These are not class-based modules. this is the simplest way I've found. we use local storage only for developing purposes. So I thought this is enough for now because there is only one position we want to change.

File: src/firebase.ts (hunk 16 above)
Comment (amevck) [reply src/firebase.ts:6]: It's not a problem. I'll do this as a best practice. in most cases, these configurations are included in index.html. there is firebase rules for whitelist some domains. only whiteList domains will access data (we can control the access based on the domain).

File: src/components/grid/GridLayout/GridLayout.tsx (hunk 12 above)
Comment (man-csl) [reply src/components/grid/GridLayout/GridLayout.tsx:1]: please add comments in the code related to these.

File: src/components/widgetCRUD/services/widgetConfService.ts (hunk 15 above)
Comment (man-csl) [reply src/components/widgetCRUD/services/widgetConfService.ts:5]: is using class based modules discouraged? Even in that case there should be good ways to implement this type of functionality.

File: src/firebase.ts (hunk 16 above)
Comment (man-csl) [reply src/firebase.ts:6]: ok, can you please point me to some documentation from google related to this?

Comment (amevck) [reply src/firebase.ts:6]: https://stackoverflow.com/questions/37482366/is-it-safe-to-expose-firebase-apikey-to-the-public

File: src/components/widgetCRUD/services/widgetConfService.ts (hunk 15 above)
Comment (amevck) [reply src/components/widgetCRUD/services/widgetConfService.ts:5]: Unfortunately, I couldn't find a good way. However, I can implement the class-based approach?

File: src/firebase.ts (hunk 16 above)
Comment (amevck) [reply src/firebase.ts:6]: and lot of firebase examples they put these keys to index.html

Comment (amevck) [reply src/firebase.ts:6]: this is an example project in the firebase site.  https://github.com/FirebaseExtended/firechat/blob/master/website/index.html

check line 194

Comment (man-csl) [reply src/firebase.ts:6]: So apparently this key is only used by firebase to find the api, not for authentication. Can you let me know how your API is protected?

Comment (amevck) [reply src/firebase.ts:6]: since we only use firebase as a side database (Firestore) we can't handle authentication things. we can manage authorization using firebase security rules (to some level). [https://firebase.google.com/docs/rules](url)

Comment (man-csl) [reply src/firebase.ts:6]: strange. how does it allow protecting different user's resources who are in the same domain?

Comment (amevck) [reply src/firebase.ts:6]: we can manage if the authentication is also in the firebase. but the scenario is different for our case.

Comment (man-csl) [reply src/firebase.ts:6]: As discussed, authentication happens through some external auth provider such as google/azure and in firebase we need to provide authorization (Access control). We usually get an access key during authentication and need to find how we can use that to protect the firebase API.

File: src/components/widgetCRUD/services/widgetConfService.ts (hunk 15 above)
Comment (man-csl) [reply src/components/widgetCRUD/services/widgetConfService.ts:5]: Lets create a task to research/work on abstracting out config saving mechanism and close this PR.

File: src/firebase.ts (hunk 16 above)
Comment (man-csl) [reply src/firebase.ts:6]: As discussed, Lets research and implement auth side of the firebase access possibly using firebase's manual auth flow + ID token returned by cognite SDK under a new task.

**Review by man-csl — APPROVED** [review]
Will be creating two new tasks to cover two unresolved review comments.
